### ISSUE TITLE
non-concurrent cache  (question)

### ISSUE THREAD

**Issue by bobymicroby** [ben-manes/caffeine#38]
Hello @ben-manes,

I would like to say that from a bird's eye your work looks amazing. Nice benchmarks/results, very good test suite and code quality. I haven't dig up that deeply, but from what I see there is no non-concurrent version of the cache. Do you have intentions to implement such version or you are going to maintain only a concurrent version ( which may be lock free, but still, it is neither wait free, nor it will have the same latency characteristics of a data structure with all the concurrency related code removed )  ?

**Comment by ben-manes** [comment]
I'm not planning on providing a non-concurrent version, can you explain why that would be desirable?

Ideally the single threaded performance shouldn't be that much worse than a non-concurrent version, despite the added complexity required. I quickly hacked the [GetPutBenchmark](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/jmh/java/com/github/benmanes/caffeine/cache/GetPutBenchmark.java) to show the results below. The upper bound should be an unsynchronized LinkedHashMap and the performance loss is reasonable when compared to other concurrent caches. Unless configured to not use a ForkJoinPool.commonPool() thread for maintenance work, the impact is negligible.

I'd argue that in a single-threaded case the cache probably won't be the bottleneck. I'd lean towards saying the addition isn't worth the conceptual weight of adding another configuration option. Probably users would be adequately served by a LinkedHashMap (in LRU mode) with the Java 8 compute methods. Extra features like expiration, weak/soft references, etc. likely aren't of much interest there.

The primary use case I know of for a single-threaded version is very Google specific. Guava provides a trimmed down cache that is compatible with GWT for transpiling into JavaScript. Similar might be offered for J2ObjC transpiling to Objective-C (iOS).

Note that the concurrent cache is not lock-free. Rather it avoids lock contention and unnecessary memory barriers from delaying the cpu.

|Policy|Read|Write
|---|---|---|
LinkedHashMap_Lru (synchronized) | 24,585,737 | 25,302,704
LinkedHashMap_Lru (unsynchronized) | 27,493,881 | 27,553,367
Caffeine (default - FJP cleanup) | 27,947,147 | 12,990,017
Caffeine (same thread cleanup) | 11,626,684 | 8,832,595
Guava | 9,112.000 | 8,017,133
Ehcache2_Lru | 6,913,574 | 2,802,196
Ehcache3_Lru | 6,146,850 | 3,702,588

**Comment by bobymicroby** [comment]
It will be interesting to include  Martin Thompson's non-concurrent, open addressing lru [Int2ObjectCache](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/collections/Int2ObjectCache.java)  to the benchmarked implementations.

**Comment by ben-manes** [comment]
Its an n-way associative cache, which I think is a better way of describing it.

I did try adding it to the simulator when he publicized it a few weeks ago (note that its a month old, not 1yr like copyright implies). I think it has similar LRU hit rates. At first I accidentally grabbed his [IntLruCache](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/collections/IntLruCache.java) which has uncharacteristically bad performance.

**Comment by bobymicroby** [comment]
I think IntLruCache was written by Richard Warburton, and yes, your definition of the data structure is more complete. I must do some benchmarks definitely.  Me and my friends are trying to write a matching engine with good latency profile. We are pretty much trying to use single threaded code, cpu pinned. We are trying to avoid concurrent code at all (  accept for the message parsing between threads ). This is why I was wondering about your opinion.

**Comment by ben-manes** [comment]
Oh right. My complaint is that his version does an O(n) move on every access. I'm not sure why he didn't use `System.arrayCopy` at least. It definitely isn't very performant.

I'm not going to add Argona into the suite because its unfair to them. That benchmark requires concurrent access with Object keys which negates their advantage.

|Benchmark|Read|Write
|---|---|---|
int | 126,676,062 | 27,466,359
Integer | 113,786,703 | 26,011,716
Integer + synchronized | 44,033,221 | 25,560,991

Whenever you can specialize a data structure for your domain its a great performance win. If you can specialize to integers then the best approach is probably an n-way cache with TinyLFU filter and a custom API. It looks like you could borrow his `IntLruCache` and my `FrequencySketch` to combine them easily enough.

My gut feeling is that its not appropriate for Caffeine to provide this all. But it is good to advocate the ideas so that others can write specialized versions or provide them in a fastutil style library.

**Comment by bobymicroby** [comment]
Thanks for the answers and your time. Wish you nice weekend :)

**Comment by ben-manes** [comment]
Have a great weekend too! Thanks for the feedback, its always welcome. :)

**Comment by ben-manes** [comment]
I took another look at `Int2ObjectCache` and one thing to be aware of is that it uses a very poor quality hash function. This makes a lot of sense, as I saw similar performance and a drop when making the hash function for the `FrequencySketch` more robust.

They took the hash function from Java 8's ConcurrentHashMap, though removed the comments explaining why a poor spreader was not a concern (degrades into a tree). For an n-way cache the degredation is a higher miss rate with a higher throughput due to cpu cache efficiencies. The simpler function results in fewer data dependencies in the pipeline, so overall a greater throughput. The clustering effect is very noticeable in Ehcache 3 which is one reason why its random sampled eviction to be very slow.

The sketch uses a better hash function, thanks to advise from @thomasmueller. It also includes a random seed to protect against collision attacks. I suspect that in some traces `Int2ObjectCache` would have a poor hit rate but that could be compensated by a TinyLFU filter. Of course that should use a higher quality hash function, but the hash could be reused between the two structures.

In short, as an n-way associative table the performance is dominated by the hash function (just like my sketch). The lower quality the higher throughput, with a probably negative impact to the hit rate and resiliency to some workloads.

**Comment by bobymicroby** [comment]
Once again, thanks again for spending so much time and for the detailed summary.
